unity-client/Assets/Rendering/Culling/CullingController.cs

```diff
@@ -0,0 +1,551 @@
+
+                //NOTE(Brian): Need to retrieve positions every frame to take into account
+                //             world repositioning.
+                Vector3 playerPosition = CommonScriptableObjects.playerUnityPosition;
+
+                Bounds bounds = r.bounds;
+                Vector3 boundingPoint = bounds.ClosestPoint(playerPosition);
+                float distance = Vector3.Distance(playerPosition, boundingPoint);
+                bool boundsContainsPlayer = bounds.Contains(playerPosition);
+                float boundsSize = bounds.size.magnitude;
+                float viewportSize = (boundsSize / distance) * Mathf.Rad2Deg;
+
+                bool isEmissive = IsEmissive(r);
+                bool isOpaque = IsOpaque(r);
+
+                float shadowTexelSize = ComputeShadowMapTexelSize(boundsSize, urpAsset.shadowDistance, urpAsset.mainLightShadowmapResolution);
+
+                bool shouldBeVisible = ShouldBeVisible(profile, viewportSize, distance, boundsContainsPlayer, isOpaque, isEmissive);
+                bool shouldHaveShadow = ShouldHaveShadow(profile, viewportSize, distance, shadowTexelSize);
+
+                SetCullingForRenderer(r, shouldBeVisible, shouldHaveShadow);
+
+                if (OnDataReport != null)
+                {
+                    if (!shouldBeVisible && !hiddenRenderers.Contains(r))
+                        hiddenRenderers.Add(r);
+
+                    if (shouldBeVisible && !shouldHaveShadow && !shadowlessRenderers.Contains(r))
+                        shadowlessRenderers.Add(r);
+                }
+
+                if (r is SkinnedMeshRenderer skr)
+                {
+                    skr.updateWhenOffscreen = ShouldUpdateSkinnedWhenOffscreen(settings, distance);
+                }
+#if UNITY_EDITOR
+                DrawDebugGizmos(shouldBeVisible, bounds, boundingPoint);
+#endif
+                timeBudgetCount += Time.realtimeSinceStartup - startTime;
+            }
+        }
+
+        /// <summary>
+        /// Main culling loop. Controlled by Start() and Stop() methods. 
+        /// </summary>
+        IEnumerator UpdateCoroutine()
+        {
+            RaiseDataReport();
+            profiles = new List<CullingControllerProfile> {settings.rendererProfile, settings.skinnedRendererProfile};
+            CommonScriptableObjects.rendererState.OnChange += (current, previous) => SetDirty();
+
+            while (true)
+            {
+                if (!CommonScriptableObjects.rendererState.Get())
+                {
+                    timeBudgetCount = 0;
+                    yield return null;
+                    continue;
+                }
+
+                bool shouldCheck = objectsTracker.IsDirty();
+
+                yield return objectsTracker.PopulateRenderersList();
+
+                Vector3 playerPosition = CommonScriptableObjects.playerUnityPosition;
+
+                if (Vector3.Distance(playerPosition, lastPlayerPos) > 1.0f)
```

Comment:
We could also subscribe to the `CommonScriptableObjects.playerUnityPosition` `OnChange` event and raise a dirty flag  or something when that event is triggered, instead of comparing vectors here.

Reply:
now all the logic is fully event driven